src/system_data.jl

```diff
@@ -1072,45 +1072,34 @@ end
 get_components_by_name(::Type{T}, data::SystemData, args...) where {T} =
     get_components_by_name(T, data.components, args...)
 
-function get_associated_components(data::SystemData, attribute::SupplementalAttribute)
-    [
-        get_component(data, x) for x in list_associated_component_uuids(
-            data.supplemental_attribute_manager.associations,
-            attribute,
-        )
-    ]
-end
-
-get_associated_components(
-    filter_func::Function,
-    data::SystemData,
-    attribute::SupplementalAttribute,
-) =
-    filter(filter_func, get_associated_components(data, attribute))
-
 function get_associated_components(
     data::SystemData,
-    attribute_type::Type{<:SupplementalAttribute},
+    attribute_type::Type{<:SupplementalAttribute};
+    component_type::Union{Nothing, Type{<:InfrastructureSystemsComponent}} = nothing,
 )
     return [
         get_component(data, x) for x in
-        list_associated_component_uuids(data.supplemental_attribute_manager, attribute_type)
+        list_associated_component_uuids(
+            data.supplemental_attribute_manager,
+            attribute_type,
+            component_type,
+        )
     ]
 end
 
 """
 Return all components associated with the attribute that match `component_type`.
 
 # Arguments
-- `component_type::{Type::<:InfrastructureSystemsComponent}`: Required type of the
-components to return. Can be concrete or abstract.
 - `data::SystemData`: the `SystemData` to search
 - `attribute::SupplementalAttribute`: Only return components associated with this attribute.
+- `component_type::Union{Nothing, Type{<:InfrastructureSystemsComponent}}`: Optional, type of the
+  components to return. Can be concrete or abstract.
 """
 function get_associated_components(
-    component_type::Type{<:InfrastructureSystemsComponent},
     data::SystemData,
-    attribute::SupplementalAttribute,
+    attribute::SupplementalAttribute;
+    component_type::Union{Nothing, Type{<:InfrastructureSystemsComponent}} = nothing,
```

Comment:
Note this change. I did mention it yesterday, but I'm not sure it completely registered with you. I think this is the best interface, but it is different than the `get_components` interface. `component_type` is an optional argument placed last. There is not a separate method with component_type listed first.

I think this is ok because `get_components` requires the component type argument. But, there could be a debate here.